packages/flutter_tools/lib/src/test/web_test_compiler.dart

```diff
@@ -0,0 +1,152 @@
+      generatedFile
+        ..createSync(recursive: true)
+        ..writeAsStringSync(generateTestEntrypoint(
+            relativeTestPath: relativeTestSegments.join('/'),
+            absolutePath: testFilePath,
+            languageVersion: languageVersion,
+        ));
+      generatedFiles.add(generatedFile);
+    }
+    // Generate a fake main file that imports all tests to be executed. This will force
+    // each of them to be compiled.
+    final StringBuffer buffer = StringBuffer('// @dart=${languageVersion.major}.${languageVersion.minor}\n');
+    for (final File generatedFile in generatedFiles) {
+      buffer.writeln('import "${_fileSystem.path.basename(generatedFile.path)}";');
+    }
+    buffer.writeln('void main() {}');
+    _fileSystem.file(_fileSystem.path.join(outputDirectory.path, 'main.dart'))
+      ..createSync()
+      ..writeAsStringSync(buffer.toString());
+
+    final String cachedKernelPath = getDefaultCachedKernelPath(
+      trackWidgetCreation: buildInfo.trackWidgetCreation,
+      dartDefines: buildInfo.dartDefines,
+      extraFrontEndOptions: extraFrontEndOptions,
+      fileSystem: _fileSystem,
+      config: _config,
+    );
+    final ResidentCompiler residentCompiler = ResidentCompiler(
+      _artifacts.getArtifactPath(Artifact.flutterWebSdk, mode: buildInfo.mode),
+      buildMode: buildInfo.mode,
+      trackWidgetCreation: buildInfo.trackWidgetCreation,
+      fileSystemRoots: <String>[
+        projectDirectory.childDirectory('test').path,
+        testOutputDir,
+      ],
+      // Override the filesystem scheme so that the frontend_server can find
+      // the generated entrypoint code.
+      fileSystemScheme: 'org-dartlang-app',
+      initializeFromDill: cachedKernelPath,
+      targetModel: TargetModel.dartdevc,
+      extraFrontEndOptions: extraFrontEndOptions,
+      platformDill: _fileSystem.file(_artifacts
+        .getArtifactPath(platformDillArtifact, mode: buildInfo.mode))
+        .absolute.uri.toString(),
+      dartDefines: buildInfo.dartDefines,
+      librariesSpec: _fileSystem.file(_artifacts
+        .getArtifactPath(Artifact.flutterWebLibrariesJson)).uri.toString(),
+      packagesPath: buildInfo.packagesPath,
+      artifacts: _artifacts,
+      processManager: _processManager,
+      logger: _logger,
+      platform: _platform,
+    );
+
+    final CompilerOutput output = await residentCompiler.recompile(
+      Uri.parse('org-dartlang-app:///main.dart'),
+      <Uri>[],
+      outputPath: outputDirectory.childFile('out').path,
+      packageConfig: buildInfo.packageConfig,
+    );
+    await residentCompiler.shutdown();
+    if (output.errorCount > 0) {
+      throwToolExit('Failed to compile');
+    }
+    // Cache the output kernel file to speed up subsequent compiles.
+    _fileSystem.file(cachedKernelPath).parent.createSync(recursive: true);
+    _fileSystem.file(output.outputFilename).copySync(cachedKernelPath);
```

Comment:
This is new, make sure to cache the dill to slightly speed up subsequent runs

Reply:
Are the new parts testable?

Reply:
Just as tested as they were before